src/random-shortcut-oriented.cc

```diff
@@ -0,0 +1,207 @@
+      RandomShortcut (problem)
+    {
+    }
+
+    PathVectorPtr_t RandomShortcutOriented::optimize (const PathVectorPtr_t& path)
+    {
+      using std::numeric_limits;
+      using std::make_pair;
+      bool finished = false;
+      value_type t0, t3;
+      const Configuration_t q0 = path->initial ();
+      const Configuration_t q3 = path->end ();
+      PathVectorPtr_t tmpPath = path;
+
+      // Maximal number of iterations without improvements
+      const std::size_t n = problem().getParameter(std::string("RandomShortcutOriented/PathOptimizersNumberOfLoops")).intValue();//, 5);
+      std::size_t projectionError = n;
+      std::deque <value_type> length (n-1,
+                                      numeric_limits <value_type>::infinity ());
+      length.push_back (PathLength<>::run (tmpPath, problem ().distance ()));
+      PathVectorPtr_t result;
+      Configuration_t q1 (path->outputSize ()),
+          q2 (path->outputSize ());
+
+      while (!finished && projectionError != 0) {
+        t0 = tmpPath->timeRange ().first;
+        t3 = tmpPath->timeRange ().second;
+        value_type u2 = t0 + (t3 -t0) * rand ()/RAND_MAX;
+        value_type u1 = t0 + (t3 -t0) * rand ()/RAND_MAX;
+        value_type t1, t2;
+        if (u1 < u2) {t1 = u1; t2 = u2;} else {t1 = u2; t2 = u1;}
+        if (!(*tmpPath) (q1, t1)) {
+          hppDout (error, "Configuration at param " << t1 << " could not be "
+                   "projected");
+          projectionError--;
+          continue;
+        }
+        if (!(*tmpPath) (q2, t2)) {
+          hppDout (error, "Configuration at param " << t1 << " could not be "
+                   "projected");
+          projectionError--;
+          continue;
+        }
+        // Validate sub parts
+        bool valid [3];
+        PathPtr_t straight [3];
+        PathPtr_t oriented;
+        KinodynamicPathPtr_t castedPath;
+        bool orientedValid(false);
+        straight [0] = steer (q0, q1);
+        straight [1] = steer (q1, q2);
+        straight [2] = steer (q2, q3);
+        PathPtr_t proj [3];
+        for (unsigned i=0; i<3; ++i) {
+          PathPtr_t validPart;
+          PathValidationReportPtr_t report;
+          if (!straight [i]) valid[i] = false;
+          else {
+            if (problem().pathProjector()) {
+              valid[i] = problem().pathProjector()->apply(straight[i], proj[i]);
+              if (!valid[i]) continue;
+            } else proj[i] = straight[i];
+            valid [i] = problem ().pathValidation ()->validate(proj [i], false, validPart, report);
+            if(valid[i]){
+              castedPath = boost::dynamic_pointer_cast<KinodynamicPath>(straight[i]);
+              if(castedPath){
+                oriented = KinodynamicOrientedPath::createCopy(castedPath);
+                orientedValid = problem ().pathValidation ()->validate(oriented, false, validPart, report);
+                if(orientedValid)
+                  proj[i] = oriented;
+              }
```

Comment:
I tried to find the differences between `RandomShorcut` and this class. Is the only difference here ? I do not understand why the original class was not sufficient. Could you add a comment ?

I would also suggest to inherits from `RandomShortcut` to reuse the functions which do not change.

@pFernbach may help on this.

Reply:
Currently yes, the only difference is here but I am planning to add more changes in the future. 
The goal of this optimization method is to produce a path where the orientation of the root try to match the direction of the motion. But because this class only work when the selected Steering-method produce a `KinodynamicPath` it may be a better idea to do this in the `finishSolve()` of the kinodynamic path-planner class.

Reply:
What is the difference between the two calls to pathValidation->validate ?

Reply:
The two paths `proj` and `oriented` don't have the same orientation for the root, so `oriented` may be invalid while `proj` is valid.

Reply:
And the oriented path is computed by the method `KinodynamicOrientedPath::createCopy` ? If yes, would you mind renaming it ?

Reply:
Done in https://github.com/humanoid-path-planner/hpp-core/pull/117/commits/746bbd88d5a98b49dc06b4095c051a28b9d24d6e

For the `randomShortcutOriented `class I decided to remove it from this PR as it's  still work in progress and not generic.

Reply:
So the file `src/random-shortcut-oriented.cc` should be removed, shouldn't it ?

Reply:
Yes, it was done in https://github.com/humanoid-path-planner/hpp-core/pull/117/commits/0bceb3c6f5b477b2fb33485ce50a7bf219b6526a